tvb_contrib/tvb/contrib/cosimulation/cosim_history.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+#
+#
+#  TheVirtualBrain-Contributors Package. This package holds simulator extensions.
+#  See also http://www.thevirtualbrain.org
+#
+# (c) 2012-2020, Baycrest Centre for Geriatric Care ("Baycrest") and others
+#
+# This program is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License along with this
+# program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as follows:
+#
+#   Paula Sanz Leon, Stuart A. Knock, M. Marmaduke Woodman, Lia Domide,
+#   Jochen Mersmann, Anthony R. McIntosh, Viktor Jirsa (2013)
+#       The Virtual Brain: a simulator of primate brain network dynamics.
+#   Frontiers in Neuroinformatics (7:10. doi: 10.3389/fninf.2013.00010)
+
+"""
+.. moduleauthor:: Lionel Kusch <lkusch@thevirtualbrain.org>
+.. moduleauthor:: Dionysios Perdikis <dionperd@gmail.com>
+"""
+
+import numpy
+
+from tvb.simulator.history import BaseHistory
+from tvb.simulator.descriptors import Dim, NDArray
+
+
+class CosimHistory(BaseHistory):
+
+    """Class for cosimulation history implementation.
+       It stores the whole TVB state for the co-simulation synchronization time.
+       The synchronization time has to be shorter than the maximum delay
+       (it usually is equal to the minimum delay of coupling between the co-simulators).
+       It is a DenseHistory since the whole state has to be stored for all delays."""
+
+    n_time, n_var, n_node, n_mode = Dim(), Dim(), Dim(), Dim()
+
+    # Overwritting attributes not needed here:
+    weights = None
+    delays = None
+    cvars = None
+
+    buffer = NDArray(('n_time', 'n_var', 'n_node', 'n_mode'), float, read_only=False)
+
+    @property
+    def nbytes(self):
+        return self.buffer.nbytes
+
+    def __init__(self,  n_time, n_var, n_node, n_mode):
+        self.n_time = n_time  # state buffer has n past steps and the current one
+        self.n_var = n_var
+        self.n_node = n_node
+        self.n_mode = n_mode
+        self.buffer[:] = numpy.NAN
```

Comment:
why NaNs here? is there an assert somewhere else that ensures that the buffer is completely and correctly filled?

Reply:
The initialization of the NAN values is for tracking the possibility of bugs for the calculation of the coupling by the users before the start of the simulation.
There are not an assert somewhere but it can be add in the function  _update_cosim_history of the cosimulator to validate the update  of the simulator history.
Do you want to add it?

Reply:
It's a good idea, but yes please add the check.  Ideally, there would be an exception specific to the condition, 
```python
class NumericalInstability(ValueError):
    pass
```
which you can raise,
```
if nan_detected(): 
    raise NumericalInstability
```
During testing or parameter tuning process, the caller can catch this specific exception and handle it more intelligently (like reducing coupling or time step), than just aborting completely.

Reply:
I add the assert
I saw, there is a special class for TVB exception : TVBException
Want you to inheritance from this class for the different type of exception in the co-simulation class?